internal/servicelog/formatter.go

```diff
@@ -151,20 +158,70 @@ func (w *TrimWriter) Write(p []byte) (int, error) {
 	}
 }
 
-func (w *TrimWriter) write(p []byte) error {
+func (w *TrimWriter) write(line []byte) error {
+	const maxDetectFailures = 10
+	if w.layout == "" && w.detectFailures < maxDetectFailures {
+		w.layout = detectLayout(line)
+		if w.layout == "" {
+			w.detectFailures++
+		}
+	}
+
 	start := 0
-	loc := w.re.FindIndex(p)
-	if loc != nil {
-		start = loc[1]
+	if w.layout != "" {
+		prefix := timeLayoutPrefix(w.layout, line)
+
+		// TODO: do we want to just trust the field splitting and not try to
+		// parse every prefix?  This could potentially cause us to discard
+		// non-timestamp content from log lines.
+		_, err := time.Parse(w.layout, prefix)
+		if err == nil {
+			start = len(prefix)
+		} else {
+			// TODO: should this be an error?
+		}
 	}
 
-	p = p[start:]
-	for len(p) > 0 {
-		n, err := w.dest.Write(p)
-		p = p[n:]
+	line = line[start:]
+	for len(line) > 0 {
+		n, err := w.dest.Write(line)
+		line = line[n:]
 		if err != nil {
 			return err
 		}
 	}
 	return nil
 }
+
+func timeLayoutPrefix(layout string, line []byte) string {
+	if layout == "" {
+		return ""
+	}
+
+	nfields := len(strings.Split(layout, " "))
```

Comment:
For efficiency, this could be done only when the layout changes.

Reply:
Yeah - I can just cache the number of fields.

Reply:
This is done.